숫자게임 [STEP1] unchain123, KayAhnDS

unchain123 opened the pull request:
@kimkyunghun3 
🔥 전역변수

initialComputerInput ← 컴퓨터 숫자들 위한 중복 방지용 세트

verifiedComputerArray ← 세트를 인덱스로 접근해 비교하기 위한 배열 (컴퓨터)

verifiedUserArray ← 세트를 인덱스로 접근해 비교하기 위한 배열 (사용자)

remainingChance ← 사용자의 총 기회

strikeFlag ← 3 스트라이크를 판단하기 위한 플레그

🔥 함수

generateRandomNumbersForComputer() → 1부터 9사이의 임의의 숫자들을 컴퓨터의 배열에 넣어주는 함수.

generateRandomNumbersForUser() → 1부터 9사이의 임의의 숫자들을 사용자의 배열에 넣어주는 함수.

decide() → 스트라이크와 볼 카운트 및 출력 함수.

gameStart() → 메인 함수로서 게임을 시작하기 위한 함수.

🔥 고민 및 해결한 점 - 코드

1️⃣ 컴퓨터의 숫자들과 사용자의 숫자들을 무작위로 받을때 어떻게 중복을 해결할 수 있을까 고민하다가 Set 컬렉션 타입을 사용하고 각각의 Set 변수의 총 요소의 수가 정해놓은 일정 수준에 도달하게 되면 반복문을 종료하고 다시 각각 컴퓨터와 사용자의 배열에 강제 형변환으로 넣어 주었습니다.

2️⃣ “컴퓨터는 한번 값들을 받으면 게임이 끝날때 까지 고정이지만, 사용자는 매번 바뀌기 때문에 어떻게 처리를 해주어야 할까” 고민하다, 각각의 함수를 따로 만들기로 결론이 났고 그에 따라 사용자를 위한 Set변수를 매번 선언하는 방식으로 사용과 초기화를 동시에 해결 했습니다. 

3️⃣ 게임이 시작되고 만약 스트라이크가 나오는 상황에서 어떻게 while문을 나올지 고민한 결과 Bool 타입의 변수를 전역변수로 선언과 함께 false로 초기화 해주고 스트라이크인 경우, Bool 타입 변수를 true로 바꿔줬습니다. 이에따라 스트라이크 및 볼을 결정 및 출력하는 함수 decide()에서 나오자마자 Bool 타입 변수의 값에 따라 반복문을 빠져 나오도록 했습니다.

4️⃣ decide() → 스트라이크 및 볼 판단 함수에서 어떻게 하면 효율적으로 볼과 스트라이크를 판단할 수 있을까 고민하던중, 먼저 스트라이크를 세는 변수와 볼을 세는 변수를 각각 만들어 주었습니다.  반복문을 돌면서 각각의 배열의 동일한 인덱스에 접근하여, 만약 두 값이 같다면 스트라이크를 하나 추가하고, 그게 아니라면 컴퓨터의 배열 내 인덱스로 접근한 값이 사용자의 배열에 포함이 된다면 볼 카운트를 하나씩 올려주는 방식으로 해결 했습니다.

5️⃣ 3 스트라이크로 게임이 끝나는 경우를 decide()내에서 어떻게 처리할까 고민했습니다. 먼저 함수 내 반복문이 끝나면 스트라이크 카운트 값이 3과 같다면 지정 문자열과 사용자 승리 문구를 함께 출력하고, 만약 3과 같지 않다면 해당 스트라이크와 볼로 출력했습니다.

6️⃣ 게임을 시작하는 함수 startGame()에서 사용자의 숫자를  한칸씩 띄워 출력하는 부분을 고민했습니다. 고민한 결과, 먼저 String 타입의 변수를 하나 생성해주고 배열의 요소와 빈칸을 함께 변수에 추가해주는 형식으로 했습니다. 이 결과 마지막에 한칸의 공백이 생기는 부분은 String 타입의 내장함수 removeLast()로 해결 해주고 변수를 그대로 출력하는 방법으로 문제를 해결 했습니다.

🔥 코드 이외의 궁금한 점

1️⃣ 서로가 페어 프로그래밍이 처음이고 어떻게 해야 더 효율적으로 공부하고 도움을 줄 수있는지 궁금했습니다. 처음부터 이리저리 부딪히면서 배우는 것도 좋지만 저희 둘 다 좋은 습관을 가지고 시작하고 싶은 마음입니다. 이런 부분에 도움을 많이 주셨으면 좋겠습니다!

2️⃣ 상대방이 코드를 작성할 동안 다른 상대방이 해당 파일에 다른것을 작성한 상태로 pull을 했더니 정상적으로 pull을 받을 수 없던 경우가 2번정도 있었습니다. 이에 대해 저희끼리 먼저 찾아본 결과 정확히 어떤 이유에서 그런 상황이 발생하는지 알 수 없었고 이에 대해서 자세하게 알려주시면 감사하겠습니다.

unchain123 committed:
feat : make function generateRandomNumbers

KayAhn0126 committed:
feat: make generateRandomNumbersForUser function

unchain123 committed:
feat: make function dicide

KayAhn0126 committed:
feat: make gameStart function

KayAhn0126 committed:
chore: fix minor grammar issues

KayAhn0126 committed:
chore: fix minor grammar issue

unchain123 committed:
Update README.md

kimkyunghun3 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,71 @@
+var initialComputerInput: Set<Int> = Set<Int>()
+var verifiedComputerArray: Array<Int> = Array<Int>()
+var verifiedUserArray: Array<Int> = Array<Int>()
 
```
전역 변수는 요구사항에 2가지만 만들도록 되어있습니다. 
전역 변수를 많이 사용하면 메모리 성능적으로 손해를 볼 수 있습니다(이는 추후 OOP에서 배우실 것이에요!)
요구사항 외에 전역을 없애고 내부에서 필요한 곳만 사용하도록 수정해보실까요?
(힌트: 메서드 간 소통하기 위해 매개변수, 리턴을 활용 할 수 있어요!)

unchain123 replied:
말씀하신대로 요구사항에 따라 전역변수를 2가지로 만들고 그 전역변수들을 이용하여 프로그램이 동작하도록 수정했습니다.

kimkyunghun3 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,71 @@
+var remainingChance = 9
+var strikeFlag: Bool = false
+
+func generateRandomNumbersForComputer() {
```
먼저 자세하게 메서드 명을 적어서 어떤 기능을 하는지 파악하기 쉬웠습니다! 
generate라는 말이 조금 저는 어색해보이는데 두 강의 보고 한번 바꿔보실까요?
[영어 네이밍짓는법](https://tv.naver.com/v/4980432/list/267189)
[API Design Guidelines](https://www.swift.org/documentation/api-design-guidelines/)

unchain123 replied:
generate를 get으로 바꿔 보았습니다.

kimkyunghun3 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,71 @@
+        let number: Int = Int.random(in: 1...9)
+        initialComputerInput.insert(number)
+    }
+    verifiedComputerArray = Array(initialComputerInput)
```
verifiedComputerArray 변수명에 Array를 쓰신 이유가 있을까요? 타입을 정할 때에도 배열인지 알 수 있는데 변수명에도 적은 이유가 궁금합니다😄

또한 verified라는 의미가 확인된, 입증된 의미를 말하는 것 같은데 여기에서 사용되는 컴퓨터배열은 입증된, 확인된 배열이라는 의미로 생각하신건가요? 조금 의미가 불명확해보이는데 수정해보실까요?!

unchain123 replied:
verifiedComputerArray는 extractedForComputer로 변경하였습니다.

kimkyunghun3 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,71 @@
+    verifiedUserArray = Array(initialUserInput)
+}
+
+func decide() {
```
메서드 명이 정확하게 어떤 기능을 하는지 명시해주면 좋을 것 같습니다! `decide()` 를 처음 보았을 때 메서드 명을 보고 어떤 기능을 하겠구나 추측가능한 네이밍을 해주시면 더 좋을 것같아요!
또한 현재 메서드에서 두 가지 기능을 수행하는 것 같은데 분리해보실수 있으실까요? 

2가지 기능이상을 수행하는 메서드가 존재하면 메서드 명을 짓는데 어려움을 겪을 가능성이 높습니다! (나중에 네이밍할 때 엄청난 고통이 생긴다는..)
 그러므로 분리해보는 연습을 해보시는 것을 추천합니다👍

unchain123 replied:
decide함수를 isStrikeBall 함수와 checkThreeStrikes함수로 나누었습니다.

kimkyunghun3 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,71 @@
+func decide() {
+    var strikeCount = 0
+    var ballCount = 0
+    for element in 0...2 {
```
반복문을 통해 확인하는 방법 좋은 생각인거같아요!👍

element 라는 네이밍이 처음 보는 사람이 보았을 때 어떤 것을 말하는지 불명확한 것처럼 보이네요! 네이밍 수정을 해볼까요?

unchain123 replied:
element는 index로 변경하였습니다.

kimkyunghun3 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,71 @@
+    for element in 0...2 {
+        if verifiedComputerArray[element] == verifiedUserArray[element] {
+            strikeCount += 1
+        } else if verifiedUserArray.contains(verifiedComputerArray[element]) {
+            ballCount += 1
```
contains 활용해서 ballCount 확인하는 방식 아주 좋습니다 😁

unchain123 replied:
감사합니다 😁

kimkyunghun3 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,71 @@
+    }
+}
+
+func startGame() {
```
메서드명보니 어떤 역할을 하는지 바로 파악할 수 있어서 좋았습니다 😇

이 메서드 내부에서도 2가지 이상의 기능을 수행하는 듯 싶어요!
분리를 해볼까요? 분리를 통해 이곳에서 호출만 해주는 방향도 가독성을 높일 수 있을 듯싶습니다!

unchain123 replied:
startGame함수에서 수행되던 기능들을 각각의 함수로 세분화해서 startGame에서는 함수로만 작동하도록 구현하였습니다.

kimkyunghun3 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,71 @@
+        for element in 0...2 {
+            result += "\(verifiedUserArray[element]) "
+        }
+        result.removeLast()
```
이렇게 해결하는 방식도 하나의 방법으로 괜찮다고 생각합니다! 👍
하지만 `removeLast()` 말고 배열 인덱스 활용한 출력방식으로 해보는거 어떤가요? 
배열을 쓰고 있으니 그 장점을 활용하는 것도 좋을 듯 싶어요!

unchain123 replied:
기존방식에서 배열 인덱스로 접근하여 출력하는 방법으로 변경하였습니다.

kimkyunghun3 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,71 @@
+        print(result)
+        
+        decide()
+        if strikeFlag {
+            break
```
Flag를 활용해서 해결하는 방법도 좋지만, 최대한 전역변수를 지양하는 것이 좋습니다! 다른 방식으로 한번 해결해볼까요?

또한 네이밍에서 Flag라기 보다는 명확한 네이밍을 해주시는게 다른 사람이 보았을 때 설명없이 더 이해하기 쉬울거 같다고 생각되네요 😄

unchain123 replied:
전역변수로 있었던 strikeFlage를 삭제하고 isStrikeBall함수의 리턴값으로 분기를 처리하여 반복문을 나올 수 있도록 하였습니다.

kimkyunghun3 reviewed (changes requested):
첫 프로젝트 STEP 1 하시면서 고생 많으셨습니다 :) 
처음으로 해주시고 코드도 깔끔하게 해주셔서 이해하는 데 어렵지 않았습니다!
한번에 완벽하게 하는 것은 어렵기 때문에 리뷰를 통해 고민도 해보시고 두분이 이야기 해보시며 개선점이 무엇일지 파악해보시면 좋을 듯 싶어요 👍 

🔥 고민 및 해결한 점 - 코드

1️⃣ 컴퓨터의 숫자들과 사용자의 숫자들을 무작위로 받을때 어떻게 중복을 해결할 수 있을까 고민하다가 Set 컬렉션 타입을 사용하고 각각의 Set 변수의 총 요소의 수가 정해놓은 일정 수준에 도달하게 되면 반복문을 종료하고 다시 각각 컴퓨터와 사용자의 배열에 강제 형변환으로 넣어 주었습니다.
> Set의 활용을 적절하게 한 것 같네요 👍 

2️⃣ “컴퓨터는 한번 값들을 받으면 게임이 끝날때 까지 고정이지만, 사용자는 매번 바뀌기 때문에 어떻게 처리를 해주어야 할까” 고민하다, 각각의 함수를 따로 만들기로 결론이 났고 그에 따라 사용자를 위한 Set변수를 매번 선언하는 방식으로 사용과 초기화를 동시에 해결 했습니다.
> 맞습니다! 컴퓨터와 유저의 값들은 조금 다르기 때문에 따로 만들 생각 하신거에 박수를 보냅니다 👍 

3️⃣ 게임이 시작되고 만약 스트라이크가 나오는 상황에서 어떻게 while문을 나올지 고민한 결과 Bool 타입의 변수를 전역변수로 선언과 함께 false로 초기화 해주고 스트라이크인 경우, Bool 타입 변수를 true로 바꿔줬습니다. 이에따라 스트라이크 및 볼을 결정 및 출력하는 함수 decide()에서 나오자마자 Bool 타입 변수의 값에 따라 반복문을 빠져 나오도록 했습니다.
> 생각하신 방향도 틀린 방향은 아닙니다! 하지만 요구사항에서 전역 선언 가능한 부분이 어떤 것인지 파악이 필요합니다! 코멘트에도 남겼는데 메서드간 소통을 어떻게 하면 좋을지 생각해보시고 수정 해보시면 좋을 것 같아요. 이를 통해 많이 배우실거라 장담합니다 :)

4️⃣ decide() → 스트라이크 및 볼 판단 함수에서 어떻게 하면 효율적으로 볼과 스트라이크를 판단할 수 있을까 고민하던중, 먼저 스트라이크를 세는 변수와 볼을 세는 변수를 각각 만들어 주었습니다. 반복문을 돌면서 각각의 배열의 동일한 인덱스에 접근하여, 만약 두 값이 같다면 스트라이크를 하나 추가하고, 그게 아니라면 컴퓨터의 배열 내 인덱스로 접근한 값이 사용자의 배열에 포함이 된다면 볼 카운트를 하나씩 올려주는 방식으로 해결 했습니다.
> 판단하는 로직이 맘에듭니다! contains를 활용해서 볼 카운트를 확인하는 등 배열의 장점들을 활용 잘했다고 생각합니다!

5️⃣ 3 스트라이크로 게임이 끝나는 경우를 decide()내에서 어떻게 처리할까 고민했습니다. 먼저 함수 내 반복문이 끝나면 스트라이크 카운트 값이 3과 같다면 지정 문자열과 사용자 승리 문구를 함께 출력하고, 만약 3과 같지 않다면 해당 스트라이크와 볼로 출력했습니다.
> 생각하신 방법으로 하셔도 괜찮을 것 같습니다! Strike가 기준이 되어서 승리 여부를 파악하는 게임이기 때문에 적절한 로직설계같습니다 👍 

6️⃣ 게임을 시작하는 함수 startGame()에서 사용자의 숫자를 한칸씩 띄워 출력하는 부분을 고민했습니다. 고민한 결과, 먼저 String 타입의 변수를 하나 생성해주고 배열의 요소와 빈칸을 함께 변수에 추가해주는 형식으로 했습니다. 이 결과 마지막에 한칸의 공백이 생기는 부분은 String 타입의 내장함수 removeLast()로 해결 해주고 변수를 그대로 출력하는 방법으로 문제를 해결 했습니다.
> 내장함수를 활용해서 작성한 코드 방향은 좋은 것 같습니다! 하지만 저희가 배열로 사용하고 있는데 이를 활용안하고 String 타입의 변수로 옮겨서 할 필요가 있을까 생각이 드네요! 정답은 없지만 배열을 활용해서 해결할 수 있을지에 대한 고민을 해보셨으면 좋겠습니다 😄

🔥 코드 이외의 궁금한 점

1️⃣ 서로가 페어 프로그래밍이 처음이고 어떻게 해야 더 효율적으로 공부하고 도움을 줄 수있는지 궁금했습니다. 처음부터 이리저리 부딪히면서 배우는 것도 좋지만 저희 둘 다 좋은 습관을 가지고 시작하고 싶은 마음입니다. 이런 부분에 도움을 많이 주셨으면 좋겠습니다!
> 제 생각에는 페어 프로그래밍은 사람끼리 하는 것이기에 두분이 편한 방법과 좋은 방법을 합의하에 선택하면 더 좋을듯 싶습니다! 제가 방법을 제시하더라도 그 방법이 여러분께 맞지 않을 수도 있으므로 만약 도중에 좋은 방향인지 의문이 생기셨다면 두분이 페어 프로그래밍 하면서 어떤 부분이 힘들었고 어떤 부분이 괜찮았는지 말씀해주실 수 있을까요?

2️⃣ 상대방이 코드를 작성할 동안 다른 상대방이 해당 파일에 다른것을 작성한 상태로 pull을 했더니 정상적으로 pull을 받을 수 없던 경우가 2번정도 있었습니다. 이에 대해 저희끼리 먼저 찾아본 결과 정확히 어떤 이유에서 그런 상황이 발생하는지 알 수 없었고 이에 대해서 자세하게 알려주시면 감사하겠습니다.
>  동일한 브런치에서 다른 분이 작업하고 있고 또 상대방이 다른 것을 작성한 상태에서 다른 분의 코드를 pull 받고 싶을 때 되지 않는 이유는 먼저 작성하신 코드가 있기 때문에 pull 이 정상적으로 되지 않습니다! pull이 된다면 현재 작성한 코드와 가져올 코드의 충돌이 발생할 수 있겠죠?(아닐수도 있지만 가능성이 큽니다)
그래서 만약 이 코드를 냅두고 다른 사람의 코드를 먼저 보고 싶다면 stash 라는 기능이 있는데 임시저장소로 작성하신 코드를 옮길 수 있어요! 그렇게 한 이후 다른 사람의 코드 pull 받고 apply 하는 방법이 있는데, 어려우시면 일단 stash의 기능에 대해 알아보시면 좋을 것 같습니다!
그게 아니라 필요없는 코드 작성 이였다면 그 코드를 discard하거나 실제 xcode에서 지우게 되면 정상적으로 pull이 받아질 것입니다!

### 추가 리뷰
![커밋명](https://user-images.githubusercontent.com/52434820/163101096-60ecb677-17f3-4ede-ab18-19915ddb1c8e.png)
커밋명에도 좀 더 정확한 표현을 하면 좋을 것 같습니다! 
커밋을 보았을 때 어떤 역할을 수행했는지 파악하면 좋아요! 포괄된 커밋명일수록 처음 보는 사람이 보았을 때 헷갈릴 수 있는 가능성이 존재합니다!
커밋명이 길어질 것 같다면, 커밋 description를 활용해보는 건 어떨까요?

KayAhn0126 committed:
refactor: removed strikeFlag

unchain123 committed:
refactor: renamed global variables

KayAhn0126 committed:
refactor: add and fix functions

add return type on getRandomNumbersForUser function, add printingRandomNumbers and fix startGame Function

unchain123 committed:
refactor: divide decied function into two functions

KayAhn0126 committed:
feat: added two functions

added subtracting and counting functions for global variable remainingchance

unchain123 committed:
refactor: refactored startGame function

unchain123 committed:
Update README.md

KayAhn0126 commented:
@kimkyunghun3

자세한 리뷰 감사합니다!
리뷰 해주신대로 전역변수를 두개만 사용하는 기준으로 코드를 리팩토링 하였습니다.

1. generate로 시작되는 함수들의 이름을 get으로 바꿔 주었습니다.
2.decide()함수를 세분화하고 그에 맞는 함수명으로 바꿔 주었습니다.
3.for-loop내 element를 index로 바꿔 주었습니다.
4.removeLast()내장함수를 쓰지않고 문자열 보간법을 이용해 그대로 출력하는 함수를 만들었습니다.
5.strikeFlag라는 전역변수를 쓰는대신 isStrikeBall함수와 checkThreeStrikes 함수를 통해 값을 반환하고 반환 받은 값으로 while문을 빠져 나오도록 하였습니다.
6. startGame()함수 내부를 말씀해 주신대로 메서드간 소통이 이뤄지게끔 수정하였습니다.

저희가 이번에 코드를 리팩토링 하면서 메서드간 소통과 불필요한 변수 생성을 막는것을 기준으로 삼았습니다.

그리고 알려주신대로 저희 둘만의 그라운드 룰을 확립하고 진행하니 한결 더 수월 했습니다.

unchain123 committed:
refactor: remove whitespace

unchain123 committed:
test

unchain123 committed:
Merge branch 'ic_6_unchain123' of https://github.com/unchain123/ios-number-baseball into ic_6_unchain123

unchain123 committed:
chore: remainingChance number change
